File: frontend/lib/models/exercise.dart
class Exercise {
  final int id;
  final String kind;
  final String front;
  final String back;
  final String credits;

  const Exercise({
    required this.id,
    required this.kind,
    required this.front,
    required this.back,
    required this.credits,
  });

  factory Exercise.fromJson(Map<String, dynamic> json) {
    return Exercise(
      id: json['id'] as int,
      kind: json['kind'] as String,
      front: json['front'] as String,
      back: json['back'] as String,
      credits: json['credits'] as String,
    );
  }
}

/// One entry from GET /exercises/pool/ — an Exercise plus the situations it's
/// eligible under, used to populate the local cache.
class ExercisePoolEntry {
  final Exercise exercise;
  final List<int> situationIds;

  const ExercisePoolEntry({required this.exercise, required this.situationIds});

  factory ExercisePoolEntry.fromJson(Map<String, dynamic> json) {
    return ExercisePoolEntry(
      exercise: Exercise.fromJson(json),
      situationIds: (json['situation_ids'] as List).cast<int>(),
    );
  }
}

File: vue/src/entities/exercise/exercise.ts
import { apiGet } from '@/db/api-client'

export interface Exercise {
  id: number
  kind: string
  front: string
  back: string
  credits: string
}

/** One entry from GET /exercises/pool/ — an exercise plus the situations it's
 * eligible under, used to populate the local cache. */
export interface ExercisePoolEntry {
  exercise: Exercise
  situationIds: number[]
}

interface RawPoolEntry extends Exercise {
  situation_ids: number[]
}

export async function generateExercise(situationId: number): Promise<Exercise> {
  return (await apiGet('/exercises/generate/', {
    situation_id: String(situationId),
  })) as Exercise
}

export async function fetchExercisePool(situationIds: Set<number>): Promise<ExercisePoolEntry[]> {
  const data = (await apiGet('/exercises/pool/', {
    situation_ids: [...situationIds].join(','),
  })) as { results: RawPoolEntry[] }

  return data.results.map((row) => ({
    exercise: { id: row.id, kind: row.kind, front: row.front, back: row.back, credits: row.credits },
    situationIds: row.situation_ids,
  }))
}
